reward/staking_manager.go

```diff
@@ -40,65 +41,89 @@ type blockChain interface {
 }
 
 type StakingManager struct {
-	ac           *addressBookConnector
-	sic          *stakingInfoCache
-	gh           governanceHelper
-	bc           blockChain
-	chainHeadCh  chan blockchain.ChainHeadEvent
-	chainHeadSub event.Subscription
-	isActivated  bool
+	addressBookConnector *addressBookConnector
+	stakingInfoCache     *stakingInfoCache
+	dbManager            database.DBManager // to store staking info into db
+	governanceHelper     governanceHelper
+	blockchain           blockChain
+	chainHeadChan        chan blockchain.ChainHeadEvent
+	chainHeadSub         event.Subscription
+	isActivated          bool
 }
 
-func NewStakingManager(bc blockChain, gh governanceHelper) *StakingManager {
+func NewStakingManager(bc blockChain, gh governanceHelper, dbm database.DBManager) *StakingManager {
 	return &StakingManager{
-		ac:          newAddressBookConnector(bc, gh),
-		sic:         newStakingInfoCache(),
-		gh:          gh,
-		bc:          bc,
-		chainHeadCh: make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
-		isActivated: false,
+		addressBookConnector: newAddressBookConnector(bc, gh),
+		stakingInfoCache:     newStakingInfoCache(),
+		dbManager:            dbm,
+		governanceHelper:     gh,
+		blockchain:           bc,
+		chainHeadChan:        make(chan blockchain.ChainHeadEvent, chainHeadChanSize),
+		isActivated:          false,
 	}
 }
 
 // GetStakingInfo returns a corresponding stakingInfo for a blockNum.
 func (sm *StakingManager) GetStakingInfo(blockNum uint64) *StakingInfo {
 	stakingBlockNumber := params.CalcStakingBlockNumber(blockNum)
 
-	if cachedStakingInfo := sm.sic.get(stakingBlockNumber); cachedStakingInfo != nil {
-		logger.Debug("StakingInfoCache hit.", "Block number", blockNum, "number of staking block", stakingBlockNumber)
+	// Get staking info from cache
+	if cachedStakingInfo := sm.stakingInfoCache.get(stakingBlockNumber); cachedStakingInfo != nil {
+		logger.Debug("StakingInfoCache hit.", "Block number", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", cachedStakingInfo)
 		return cachedStakingInfo
 	}
 
-	stakingInfo, err := sm.updateStakingCache(stakingBlockNumber)
+	// Get staking info from db
+	if sm.dbManager != nil {
+		if storedStakingInfo, err := sm.dbManager.ReadStakingInfo(stakingBlockNumber); storedStakingInfo != nil && err == nil {
+			logger.Debug("StakingInfoDB hit.", "Block number", blockNum, "staking block number", stakingBlockNumber, "stakingInfo", storedStakingInfo)
+			return storedStakingInfo.(*StakingInfo)
+		} else {
+			logger.Warn("Failed to get staking info from db.", "blockNum", blockNum, "err", err, "value", storedStakingInfo)
```

Comment:
If there is an error, it returns nil value?
```suggestion
			logger.Warn("Failed to get staking info from db.", "blockNum", blockNum, "err", err)
```

Reply:
Review reflected